### PR TITLE
Optimize paint

### PR DESCRIPTION
**Pull request by shmcaffrey**
Optimizes the paint and erase for the undo redo purposes, now when painting/erasing, the set of pixels is determined by traversing over a rectangle from the top left most corner the users brush touches to the bottom right most corner it touches.

### CODE REVIEW

File: src/main/ngapp/accio/src/app/editor/editor.component.ts [hunk 1]
```diff
@@ -400,7 +400,7 @@ export class EditorComponent implements OnInit {
   }
 
   /**
-   *  Makes a png of mask so the the background is transparent.
+   *  Makes a pewtopg of mask so the the background is transparent.
```
Comment (charmychancey) [inline src/main/ngapp/accio/src/app/editor/editor.component.ts:403]: Is this a typo?

File: src/main/ngapp/accio/src/app/editor/editor.component.ts [hunk 2]
```diff
@@ -931,7 +933,7 @@ class Rectangle {
     this.top = Math.max(coord.y - this.brushRadius, 0);
     this.bottom = Math.min(coord.y + this.brushRadius, this.imageHeight - 1);
 
-    this.left = Math.max(coord.x - this.brushRadius, 0);
+    this.left = Math.max(coord.x /*- this.brushRadius*/, 0);
```
Comment (charmychancey) [inline src/main/ngapp/accio/src/app/editor/editor.component.ts:936]: Is this comment still needed?

File: src/main/ngapp/accio/src/app/editor/editor.component.ts (hunk 1 above)
Comment (shmcaffrey) [reply src/main/ngapp/accio/src/app/editor/editor.component.ts:403]: I think I accidentally hit keys without knowing it oops

File: src/main/ngapp/accio/src/app/editor/editor.component.ts [hunk 3]
```diff
@@ -837,18 +835,21 @@ export class EditorComponent implements OnInit {
...
-   
     let currRightTopIndex = this.magicWandService.coordToDataArrayIndex(rightBottom.x, leftTop.y, this.image.width);
+
     let newTopY = leftTop.y;
 
-    for (let i = leftTopIndex; i < paintedImageData.length; i += 4) {
+    for (let i = leftTopIndex; i <= rightBottomIndex; i += 4) {
       // If the alpha value has value.
-      if (paintedImageData[i + 3] === 255) {
+      if (paintedImageData[i + 3] > (255 / 2)) {
```
Comment (robotsaurus) [inline src/main/ngapp/accio/src/app/editor/editor.component.ts:844]: Add a comment why you are doing this comparison.

File: src/main/ngapp/accio/src/app/editor/editor.component.ts [hunk 4]
```diff
@@ -837,18 +835,21 @@ export class EditorComponent implements OnInit {
...
         paintedMask.add(i);
       }
 
-      // TODO(SHMCAFFREY) loop through only the indicies in the rectangle.
-
+      if (i === currRightTopIndex) {
+        // Minus 4 to account for i += 4 in loop.
+        i = this.magicWandService.coordToDataArrayIndex(leftTop.x, ++newTopY, this.image.width) - 4;
+        currRightTopIndex =  this.magicWandService.coordToDataArrayIndex(rightBottom.x, newTopY, this.image.width);
+      }
```
Comment (robotsaurus) [inline src/main/ngapp/accio/src/app/editor/editor.component.ts:852]: This loop would probably be better written to loop over the (x,y) coordinates of the image and then convert to the indices which would avoid this complication/error prone logic.

No need to do this in this PR.  Just maybe add a TODO.

**Comment by shmcaffrey** [comment]
ptal

File: src/main/ngapp/accio/src/app/editor/editor.component.ts [hunk 5]
```diff
@@ -957,10 +963,22 @@ class Rectangle {
     }
   }
 
-  getLeftTop() {
-    return new Coordinate(this.left, this.top);
+ /**  
+  *  @returns {Coordinate} the floored (x,y) of the combination of the left
+  *    most and top most coordinate the users brush touched.
+  *  Floor is needed to calculate the propper index in the data array.
+  *    It allows the iindex to be a valid (x,y) coordinate
```
Comment (snazzorama) [inline src/main/ngapp/accio/src/app/editor/editor.component.ts:970]: Nit: Only one middle 'p' in 'proper' and one 'i' in index.

**Pull request merged**